// @glimmer/syntax/dist/commonjs/es5/lib/v1/legacy-interop.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PathExpressionImplV1 = void 0;

var _publicBuilders = _interopRequireDefault(require("./public-builders"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

var PathExpressionImplV1 = /*#__PURE__*/function () {
  function PathExpressionImplV1(original, head, tail, loc) {
    this.original = original;
    this.loc = loc;
    this.type = 'PathExpression';
    this["this"] = false;
    this.data = false; // Cache for the head value.

    this._head = undefined;
    var parts = tail.slice();

    if (head.type === 'ThisHead') {
      this["this"] = true;
    } else if (head.type === 'AtHead') {
      this.data = true;
      parts.unshift(head.name.slice(1));
    } else {
      parts.unshift(head.name);
    }

    this.parts = parts;
  }

  _createClass(PathExpressionImplV1, [{
    key: "head",
    get: function get() {
      if (this._head) {
        return this._head;
      }

      var firstPart;

      if (this["this"]) {
        firstPart = 'this';
      } else if (this.data) {
        firstPart = "@" + this.parts[0];
      } else {
        firstPart = this.parts[0];
      }

      var firstPartLoc = this.loc.collapse('start').sliceStartChars({
        chars: firstPart.length
      }).loc;
      return this._head = _publicBuilders.default.head(firstPart, firstPartLoc);
    }
  }, {
    key: "tail",
    get: function get() {
      return this["this"] ? this.parts : this.parts.slice(1);
    }
  }]);

  return PathExpressionImplV1;
}();

exports.PathExpressionImplV1 = PathExpressionImplV1;
