// @glimmer/syntax/dist/commonjs/es5/lib/v1/visitor-keys.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _util = require("@glimmer/util");

// ensure stays in sync with typing
// ParentNode and ChildKey types are derived from VisitorKeysMap
var visitorKeys = {
  Program: (0, _util.tuple)('body'),
  Template: (0, _util.tuple)('body'),
  Block: (0, _util.tuple)('body'),
  MustacheStatement: (0, _util.tuple)('path', 'params', 'hash'),
  BlockStatement: (0, _util.tuple)('path', 'params', 'hash', 'program', 'inverse'),
  ElementModifierStatement: (0, _util.tuple)('path', 'params', 'hash'),
  PartialStatement: (0, _util.tuple)('name', 'params', 'hash'),
  CommentStatement: (0, _util.tuple)(),
  MustacheCommentStatement: (0, _util.tuple)(),
  ElementNode: (0, _util.tuple)('attributes', 'modifiers', 'children', 'comments'),
  AttrNode: (0, _util.tuple)('value'),
  TextNode: (0, _util.tuple)(),
  ConcatStatement: (0, _util.tuple)('parts'),
  SubExpression: (0, _util.tuple)('path', 'params', 'hash'),
  PathExpression: (0, _util.tuple)(),
  PathHead: (0, _util.tuple)(),
  StringLiteral: (0, _util.tuple)(),
  BooleanLiteral: (0, _util.tuple)(),
  NumberLiteral: (0, _util.tuple)(),
  NullLiteral: (0, _util.tuple)(),
  UndefinedLiteral: (0, _util.tuple)(),
  Hash: (0, _util.tuple)('pairs'),
  HashPair: (0, _util.tuple)('value'),
  // v2 new nodes
  NamedBlock: (0, _util.tuple)('attributes', 'modifiers', 'children', 'comments'),
  SimpleElement: (0, _util.tuple)('attributes', 'modifiers', 'children', 'comments'),
  Component: (0, _util.tuple)('head', 'attributes', 'modifiers', 'children', 'comments')
};
var _default = visitorKeys;
exports.default = _default;
