// @embroider/shared-internals/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.templateColocationPlugin = exports.hbsToJS = exports.tmpdir = exports.packageName = exports.babelFilter = exports.PackageCache = exports.Package = exports.getOrCreate = exports.extensionsPattern = exports.explicitRelative = void 0;
var paths_1 = require("./paths");
Object.defineProperty(exports, "explicitRelative", { enumerable: true, get: function () { return paths_1.explicitRelative; } });
Object.defineProperty(exports, "extensionsPattern", { enumerable: true, get: function () { return paths_1.extensionsPattern; } });
var get_or_create_1 = require("./get-or-create");
Object.defineProperty(exports, "getOrCreate", { enumerable: true, get: function () { return get_or_create_1.getOrCreate; } });
var package_1 = require("./package");
Object.defineProperty(exports, "Package", { enumerable: true, get: function () { return __importDefault(package_1).default; } });
var package_cache_1 = require("./package-cache");
Object.defineProperty(exports, "PackageCache", { enumerable: true, get: function () { return __importDefault(package_cache_1).default; } });
var babel_filter_1 = require("./babel-filter");
Object.defineProperty(exports, "babelFilter", { enumerable: true, get: function () { return __importDefault(babel_filter_1).default; } });
var package_name_1 = require("./package-name");
Object.defineProperty(exports, "packageName", { enumerable: true, get: function () { return __importDefault(package_name_1).default; } });
var tmpdir_1 = require("./tmpdir");
Object.defineProperty(exports, "tmpdir", { enumerable: true, get: function () { return __importDefault(tmpdir_1).default; } });
__exportStar(require("./ember-cli-models"), exports);
__exportStar(require("./ember-standard-modules"), exports);
var hbs_to_js_1 = require("./hbs-to-js");
Object.defineProperty(exports, "hbsToJS", { enumerable: true, get: function () { return hbs_to_js_1.hbsToJS; } });
var template_colocation_plugin_1 = require("./template-colocation-plugin");
Object.defineProperty(exports, "templateColocationPlugin", { enumerable: true, get: function () { return __importDefault(template_colocation_plugin_1).default; } });
