// @glimmer/syntax/dist/commonjs/es2017/lib/v1/legacy-interop.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PathExpressionImplV1 = void 0;

var _publicBuilders = _interopRequireDefault(require("./public-builders"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class PathExpressionImplV1 {
  constructor(original, head, tail, loc) {
    this.original = original;
    this.loc = loc;
    this.type = 'PathExpression';
    this.this = false;
    this.data = false; // Cache for the head value.

    this._head = undefined;
    let parts = tail.slice();

    if (head.type === 'ThisHead') {
      this.this = true;
    } else if (head.type === 'AtHead') {
      this.data = true;
      parts.unshift(head.name.slice(1));
    } else {
      parts.unshift(head.name);
    }

    this.parts = parts;
  }

  get head() {
    if (this._head) {
      return this._head;
    }

    let firstPart;

    if (this.this) {
      firstPart = 'this';
    } else if (this.data) {
      firstPart = `@${this.parts[0]}`;
    } else {
      firstPart = this.parts[0];
    }

    let firstPartLoc = this.loc.collapse('start').sliceStartChars({
      chars: firstPart.length
    }).loc;
    return this._head = _publicBuilders.default.head(firstPart, firstPartLoc);
  }

  get tail() {
    return this.this ? this.parts : this.parts.slice(1);
  }

}

exports.PathExpressionImplV1 = PathExpressionImplV1;
