// @glimmer/syntax/dist/modules/es5/lib/v2-a/objects/attr-block.js
function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import { NamedArgument } from './args';
import { node } from './node';
/**
 * `HtmlAttr` nodes are valid HTML attributes, with or without a value.
 *
 * Exceptions:
 *
 * - `...attributes` is `SplatAttr`
 * - `@x=<value>` is `ComponentArg`
 */

export var HtmlAttr = /*#__PURE__*/function (_node$fields) {
  _inheritsLoose(HtmlAttr, _node$fields);

  function HtmlAttr() {
    return _node$fields.apply(this, arguments) || this;
  }

  return HtmlAttr;
}(node('HtmlAttr').fields());
export var SplatAttr = /*#__PURE__*/function (_node$fields2) {
  _inheritsLoose(SplatAttr, _node$fields2);

  function SplatAttr() {
    return _node$fields2.apply(this, arguments) || this;
  }

  return SplatAttr;
}(node('SplatAttr').fields());
/**
 * Corresponds to an argument passed by a component (`@x=<value>`)
 */

export var ComponentArg = /*#__PURE__*/function (_node$fields3) {
  _inheritsLoose(ComponentArg, _node$fields3);

  function ComponentArg() {
    return _node$fields3.apply(this, arguments) || this;
  }

  var _proto = ComponentArg.prototype;

  /**
   * Convert the component argument into a named argument node
   */
  _proto.toNamedArgument = function toNamedArgument() {
    return new NamedArgument({
      name: this.name,
      value: this.value
    });
  };

  return ComponentArg;
}(node().fields());
/**
 * An `ElementModifier` is just a normal call node in modifier position.
 */

export var ElementModifier = /*#__PURE__*/function (_node$fields4) {
  _inheritsLoose(ElementModifier, _node$fields4);

  function ElementModifier() {
    return _node$fields4.apply(this, arguments) || this;
  }

  return ElementModifier;
}(node('ElementModifier').fields());
