// @glimmer/syntax/dist/modules/es5/lib/v2-a/objects/content.js
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import { SpanList } from '../../source/span-list';
import { Args, NamedArguments } from './args';
import { node } from './node';
export var GlimmerComment = /*#__PURE__*/function (_node$fields) {
  _inheritsLoose(GlimmerComment, _node$fields);

  function GlimmerComment() {
    return _node$fields.apply(this, arguments) || this;
  }

  return GlimmerComment;
}(node('GlimmerComment').fields());
export var HtmlText = /*#__PURE__*/function (_node$fields2) {
  _inheritsLoose(HtmlText, _node$fields2);

  function HtmlText() {
    return _node$fields2.apply(this, arguments) || this;
  }

  return HtmlText;
}(node('HtmlText').fields());
export var HtmlComment = /*#__PURE__*/function (_node$fields3) {
  _inheritsLoose(HtmlComment, _node$fields3);

  function HtmlComment() {
    return _node$fields3.apply(this, arguments) || this;
  }

  return HtmlComment;
}(node('HtmlComment').fields());
export var AppendContent = /*#__PURE__*/function (_node$fields4) {
  _inheritsLoose(AppendContent, _node$fields4);

  function AppendContent() {
    return _node$fields4.apply(this, arguments) || this;
  }

  _createClass(AppendContent, [{
    key: "callee",
    get: function get() {
      if (this.value.type === 'Call') {
        return this.value.callee;
      } else {
        return this.value;
      }
    }
  }, {
    key: "args",
    get: function get() {
      if (this.value.type === 'Call') {
        return this.value.args;
      } else {
        return Args.empty(this.value.loc.collapse('end'));
      }
    }
  }]);

  return AppendContent;
}(node('AppendContent').fields());
export var InvokeBlock = /*#__PURE__*/function (_node$fields5) {
  _inheritsLoose(InvokeBlock, _node$fields5);

  function InvokeBlock() {
    return _node$fields5.apply(this, arguments) || this;
  }

  return InvokeBlock;
}(node('InvokeBlock').fields());
/**
 * Corresponds to a component invocation. When the content of a component invocation contains no
 * named blocks, `blocks` contains a single named block named `"default"`. When a component
 * invocation is self-closing, `blocks` is empty.
 */

export var InvokeComponent = /*#__PURE__*/function (_node$fields6) {
  _inheritsLoose(InvokeComponent, _node$fields6);

  function InvokeComponent() {
    return _node$fields6.apply(this, arguments) || this;
  }

  _createClass(InvokeComponent, [{
    key: "args",
    get: function get() {
      var entries = this.componentArgs.map(function (a) {
        return a.toNamedArgument();
      });
      return Args.named(new NamedArguments({
        loc: SpanList.range(entries, this.callee.loc.collapse('end')),
        entries: entries
      }));
    }
  }]);

  return InvokeComponent;
}(node('InvokeComponent').fields());
/**
 * Corresponds to a simple HTML element. The AST allows component arguments and modifiers to support
 * future extensions.
 */

export var SimpleElement = /*#__PURE__*/function (_node$fields7) {
  _inheritsLoose(SimpleElement, _node$fields7);

  function SimpleElement() {
    return _node$fields7.apply(this, arguments) || this;
  }

  _createClass(SimpleElement, [{
    key: "args",
    get: function get() {
      var entries = this.componentArgs.map(function (a) {
        return a.toNamedArgument();
      });
      return Args.named(new NamedArguments({
        loc: SpanList.range(entries, this.tag.loc.collapse('end')),
        entries: entries
      }));
    }
  }]);

  return SimpleElement;
}(node('SimpleElement').fields());
