// @glimmer/syntax/dist/modules/es5/lib/v2-a/objects/expr.js
function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import { SourceSlice } from '../../source/slice';
import { node } from './node';
/**
 * Corresponds to a Handlebars literal.
 *
 * @see {LiteralValue}
 */

export var LiteralExpression = /*#__PURE__*/function (_node$fields) {
  _inheritsLoose(LiteralExpression, _node$fields);

  function LiteralExpression() {
    return _node$fields.apply(this, arguments) || this;
  }

  var _proto = LiteralExpression.prototype;

  _proto.toSlice = function toSlice() {
    return new SourceSlice({
      loc: this.loc,
      chars: this.value
    });
  };

  return LiteralExpression;
}(node('Literal').fields());
/**
 * Returns true if an input {@see ExpressionNode} is a literal.
 */

export function isLiteral(node, kind) {
  if (node.type === 'Literal') {
    if (kind === undefined) {
      return true;
    } else if (kind === 'null') {
      return node.value === null;
    } else {
      return typeof node.value === kind;
    }
  } else {
    return false;
  }
}
/**
 * Corresponds to a path in expression position.
 *
 * ```hbs
 * this
 * this.x
 * @x
 * @x.y
 * x
 * x.y
 * ```
 */

export var PathExpression = /*#__PURE__*/function (_node$fields2) {
  _inheritsLoose(PathExpression, _node$fields2);

  function PathExpression() {
    return _node$fields2.apply(this, arguments) || this;
  }

  return PathExpression;
}(node('Path').fields());
/**
 * Corresponds to a parenthesized call expression.
 *
 * ```hbs
 * (x)
 * (x.y)
 * (x y)
 * (x.y z)
 * ```
 */

export var CallExpression = /*#__PURE__*/function (_node$fields3) {
  _inheritsLoose(CallExpression, _node$fields3);

  function CallExpression() {
    return _node$fields3.apply(this, arguments) || this;
  }

  return CallExpression;
}(node('Call').fields());
/**
 * Corresponds to a possible deprecated helper call. Must be:
 *
 * 1. A free variable (not this.foo, not @foo, not local).
 * 2. Argument-less.
 * 3. In a component invocation's named argument position.
 * 4. Not parenthesized (not @bar={{(helper)}}).
 * 5. Not interpolated (not @bar="{{helper}}").
 *
 * ```hbs
 * <Foo @bar={{helper}} />
 * ```
 */

export var DeprecatedCallExpression = /*#__PURE__*/function (_node$fields4) {
  _inheritsLoose(DeprecatedCallExpression, _node$fields4);

  function DeprecatedCallExpression() {
    return _node$fields4.apply(this, arguments) || this;
  }

  return DeprecatedCallExpression;
}(node('DeprecatedCall').fields());
/**
 * Corresponds to an interpolation in attribute value position.
 *
 * ```hbs
 * <a href="{{url}}.html"
 * ```
 */

export var InterpolateExpression = /*#__PURE__*/function (_node$fields5) {
  _inheritsLoose(InterpolateExpression, _node$fields5);

  function InterpolateExpression() {
    return _node$fields5.apply(this, arguments) || this;
  }

  return InterpolateExpression;
}(node('Interpolate').fields());
