// @glimmer/syntax/dist/modules/es5/lib/v2-a/objects/internal-node.js
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

import { SpanList } from '../../source/span-list';
import { Args, NamedArguments } from './args';
import { node } from './node';
/**
 * Corresponds to an entire template.
 */

export var Template = /*#__PURE__*/function (_node$fields) {
  _inheritsLoose(Template, _node$fields);

  function Template() {
    return _node$fields.apply(this, arguments) || this;
  }

  return Template;
}(node().fields());
/**
 * Represents a block. In principle this could be merged with `NamedBlock`, because all cases
 * involving blocks have at least a notional name.
 */

export var Block = /*#__PURE__*/function (_node$fields2) {
  _inheritsLoose(Block, _node$fields2);

  function Block() {
    return _node$fields2.apply(this, arguments) || this;
  }

  return Block;
}(node().fields());
/**
 * Corresponds to a collection of named blocks.
 */

export var NamedBlocks = /*#__PURE__*/function (_node$fields3) {
  _inheritsLoose(NamedBlocks, _node$fields3);

  function NamedBlocks() {
    return _node$fields3.apply(this, arguments) || this;
  }

  var _proto = NamedBlocks.prototype;

  _proto.get = function get(name) {
    return this.blocks.filter(function (block) {
      return block.name.chars === name;
    })[0] || null;
  };

  return NamedBlocks;
}(node().fields());
/**
 * Corresponds to a single named block. This is used for anonymous named blocks (`default` and
 * `else`).
 */

export var NamedBlock = /*#__PURE__*/function (_node$fields4) {
  _inheritsLoose(NamedBlock, _node$fields4);

  function NamedBlock() {
    return _node$fields4.apply(this, arguments) || this;
  }

  _createClass(NamedBlock, [{
    key: "args",
    get: function get() {
      var entries = this.componentArgs.map(function (a) {
        return a.toNamedArgument();
      });
      return Args.named(new NamedArguments({
        loc: SpanList.range(entries, this.name.loc.collapse('end')),
        entries: entries
      }));
    }
  }]);

  return NamedBlock;
}(node().fields());
