// @glimmer/syntax/dist/modules/es2017/lib/source/source.js
// eslint-disable-next-line import/no-extraneous-dependencies
import { DEBUG } from '@glimmer/env';
import { assert } from '@glimmer/util';
import { SourceOffset, SourceSpan } from './span';
export class Source {
  constructor(source, module = 'an unknown module') {
    this.source = source;
    this.module = module;
  }
  /**
   * Validate that the character offset represents a position in the source string.
   */


  check(offset) {
    return offset >= 0 && offset <= this.source.length;
  }

  slice(start, end) {
    return this.source.slice(start, end);
  }

  offsetFor(line, column) {
    return SourceOffset.forHbsPos(this, {
      line,
      column
    });
  }

  spanFor({
    start,
    end
  }) {
    return SourceSpan.forHbsLoc(this, {
      start: {
        line: start.line,
        column: start.column
      },
      end: {
        line: end.line,
        column: end.column
      }
    });
  }

  hbsPosFor(offset) {
    let seenLines = 0;
    let seenChars = 0;

    if (offset > this.source.length) {
      return null;
    }

    while (true) {
      let nextLine = this.source.indexOf('\n', seenChars);

      if (offset <= nextLine || nextLine === -1) {
        return {
          line: seenLines + 1,
          column: offset - seenChars
        };
      } else {
        seenLines += 1;
        seenChars = nextLine + 1;
      }
    }
  }

  charPosFor(position) {
    let {
      line,
      column
    } = position;
    let sourceString = this.source;
    let sourceLength = sourceString.length;
    let seenLines = 0;
    let seenChars = 0;

    while (true) {
      if (seenChars >= sourceLength) return sourceLength;
      let nextLine = this.source.indexOf('\n', seenChars);
      if (nextLine === -1) nextLine = this.source.length;

      if (seenLines === line - 1) {
        if (seenChars + column > nextLine) return nextLine;

        if (DEBUG) {
          let roundTrip = this.hbsPosFor(seenChars + column);
          (false && assert(roundTrip !== null, `the returned offset failed to round-trip`));
          (false && assert(roundTrip.line === line, `the round-tripped line didn't match the original line`));
          (false && assert(roundTrip.column === column, `the round-tripped column didn't match the original column`));
        }

        return seenChars + column;
      } else if (nextLine === -1) {
        return 0;
      } else {
        seenLines += 1;
        seenChars = nextLine + 1;
      }
    }
  }

}
