// @glimmer/syntax/dist/commonjs/es5/lib/v2-a/objects/attr-block.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ElementModifier = exports.ComponentArg = exports.SplatAttr = exports.HtmlAttr = void 0;

var _args = require("./args");

var _node = require("./node");

function _inheritsLoose(subClass, superClass) {
  subClass.prototype = Object.create(superClass.prototype);
  subClass.prototype.constructor = subClass;
  subClass.__proto__ = superClass;
}

/**
 * `HtmlAttr` nodes are valid HTML attributes, with or without a value.
 *
 * Exceptions:
 *
 * - `...attributes` is `SplatAttr`
 * - `@x=<value>` is `ComponentArg`
 */
var HtmlAttr = /*#__PURE__*/function (_node$fields) {
  _inheritsLoose(HtmlAttr, _node$fields);

  function HtmlAttr() {
    return _node$fields.apply(this, arguments) || this;
  }

  return HtmlAttr;
}((0, _node.node)('HtmlAttr').fields());

exports.HtmlAttr = HtmlAttr;

var SplatAttr = /*#__PURE__*/function (_node$fields2) {
  _inheritsLoose(SplatAttr, _node$fields2);

  function SplatAttr() {
    return _node$fields2.apply(this, arguments) || this;
  }

  return SplatAttr;
}((0, _node.node)('SplatAttr').fields());
/**
 * Corresponds to an argument passed by a component (`@x=<value>`)
 */


exports.SplatAttr = SplatAttr;

var ComponentArg = /*#__PURE__*/function (_node$fields3) {
  _inheritsLoose(ComponentArg, _node$fields3);

  function ComponentArg() {
    return _node$fields3.apply(this, arguments) || this;
  }

  var _proto = ComponentArg.prototype;
  /**
   * Convert the component argument into a named argument node
   */

  _proto.toNamedArgument = function toNamedArgument() {
    return new _args.NamedArgument({
      name: this.name,
      value: this.value
    });
  };

  return ComponentArg;
}((0, _node.node)().fields());
/**
 * An `ElementModifier` is just a normal call node in modifier position.
 */


exports.ComponentArg = ComponentArg;

var ElementModifier = /*#__PURE__*/function (_node$fields4) {
  _inheritsLoose(ElementModifier, _node$fields4);

  function ElementModifier() {
    return _node$fields4.apply(this, arguments) || this;
  }

  return ElementModifier;
}((0, _node.node)('ElementModifier').fields());

exports.ElementModifier = ElementModifier;
