// @glimmer/syntax/dist/modules/es2017/lib/v1/visitor-keys.js
import { tuple } from '@glimmer/util'; // ensure stays in sync with typing
// ParentNode and ChildKey types are derived from VisitorKeysMap

const visitorKeys = {
  Program: tuple('body'),
  Template: tuple('body'),
  Block: tuple('body'),
  MustacheStatement: tuple('path', 'params', 'hash'),
  BlockStatement: tuple('path', 'params', 'hash', 'program', 'inverse'),
  ElementModifierStatement: tuple('path', 'params', 'hash'),
  PartialStatement: tuple('name', 'params', 'hash'),
  CommentStatement: tuple(),
  MustacheCommentStatement: tuple(),
  ElementNode: tuple('attributes', 'modifiers', 'children', 'comments'),
  AttrNode: tuple('value'),
  TextNode: tuple(),
  ConcatStatement: tuple('parts'),
  SubExpression: tuple('path', 'params', 'hash'),
  PathExpression: tuple(),
  PathHead: tuple(),
  StringLiteral: tuple(),
  BooleanLiteral: tuple(),
  NumberLiteral: tuple(),
  NullLiteral: tuple(),
  UndefinedLiteral: tuple(),
  Hash: tuple('pairs'),
  HashPair: tuple('value'),
  // v2 new nodes
  NamedBlock: tuple('attributes', 'modifiers', 'children', 'comments'),
  SimpleElement: tuple('attributes', 'modifiers', 'children', 'comments'),
  Component: tuple('head', 'attributes', 'modifiers', 'children', 'comments')
};
export default visitorKeys;
