// @glimmer/syntax/dist/modules/es2017/index.js
export { Source } from './lib/source/source';
export { default as builders } from './lib/v1/public-builders';
import * as ASTv1_1 from './lib/v1/api';
export { ASTv1_1 as ASTv1 };
import * as ASTv2_1 from './lib/v2-a/api';
export { ASTv2_1 as ASTv2 };
export { normalize } from './lib/v2-a/normalize';
export { SymbolTable, BlockSymbolTable, ProgramSymbolTable } from './lib/symbol-table';
export { generateSyntaxError } from './lib/syntax-error';
export { preprocess } from './lib/parser/tokenizer-event-handlers';
export { default as print } from './lib/generation/print';
export { sortByLoc } from './lib/generation/util';
export { default as Walker } from './lib/traversal/walker';
export { default as traverse } from './lib/traversal/traverse';
export { cannotRemoveNode, cannotReplaceNode } from './lib/traversal/errors';
export { default as WalkerPath } from './lib/traversal/path';
export { isKeyword, KEYWORDS_TYPES } from './lib/keywords';
export { getTemplateLocals } from './lib/get-template-locals';
export { SourceSlice } from './lib/source/slice';
export { SourceSpan } from './lib/source/span';
export { SpanList, maybeLoc, loc, hasSpan } from './lib/source/span-list';
export { node } from './lib/v2-a/objects/node';
/** @deprecated use WalkerPath instead */

export { default as Path } from './lib/traversal/walker';
import * as AST_1 from './lib/v1/api';
export { AST_1 as AST };
