// @glimmer/syntax/dist/commonjs/es5/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Source", {
  enumerable: true,
  get: function () {
    return _source.Source;
  }
});
Object.defineProperty(exports, "builders", {
  enumerable: true,
  get: function () {
    return _publicBuilders.default;
  }
});
Object.defineProperty(exports, "normalize", {
  enumerable: true,
  get: function () {
    return _normalize.normalize;
  }
});
Object.defineProperty(exports, "SymbolTable", {
  enumerable: true,
  get: function () {
    return _symbolTable.SymbolTable;
  }
});
Object.defineProperty(exports, "BlockSymbolTable", {
  enumerable: true,
  get: function () {
    return _symbolTable.BlockSymbolTable;
  }
});
Object.defineProperty(exports, "ProgramSymbolTable", {
  enumerable: true,
  get: function () {
    return _symbolTable.ProgramSymbolTable;
  }
});
Object.defineProperty(exports, "generateSyntaxError", {
  enumerable: true,
  get: function () {
    return _syntaxError.generateSyntaxError;
  }
});
Object.defineProperty(exports, "preprocess", {
  enumerable: true,
  get: function () {
    return _tokenizerEventHandlers.preprocess;
  }
});
Object.defineProperty(exports, "print", {
  enumerable: true,
  get: function () {
    return _print.default;
  }
});
Object.defineProperty(exports, "sortByLoc", {
  enumerable: true,
  get: function () {
    return _util.sortByLoc;
  }
});
Object.defineProperty(exports, "Walker", {
  enumerable: true,
  get: function () {
    return _walker.default;
  }
});
Object.defineProperty(exports, "Path", {
  enumerable: true,
  get: function () {
    return _walker.default;
  }
});
Object.defineProperty(exports, "traverse", {
  enumerable: true,
  get: function () {
    return _traverse.default;
  }
});
Object.defineProperty(exports, "cannotRemoveNode", {
  enumerable: true,
  get: function () {
    return _errors.cannotRemoveNode;
  }
});
Object.defineProperty(exports, "cannotReplaceNode", {
  enumerable: true,
  get: function () {
    return _errors.cannotReplaceNode;
  }
});
Object.defineProperty(exports, "WalkerPath", {
  enumerable: true,
  get: function () {
    return _path.default;
  }
});
Object.defineProperty(exports, "isKeyword", {
  enumerable: true,
  get: function () {
    return _keywords.isKeyword;
  }
});
Object.defineProperty(exports, "KEYWORDS_TYPES", {
  enumerable: true,
  get: function () {
    return _keywords.KEYWORDS_TYPES;
  }
});
Object.defineProperty(exports, "getTemplateLocals", {
  enumerable: true,
  get: function () {
    return _getTemplateLocals.getTemplateLocals;
  }
});
Object.defineProperty(exports, "SourceSlice", {
  enumerable: true,
  get: function () {
    return _slice.SourceSlice;
  }
});
Object.defineProperty(exports, "SourceSpan", {
  enumerable: true,
  get: function () {
    return _span.SourceSpan;
  }
});
Object.defineProperty(exports, "SpanList", {
  enumerable: true,
  get: function () {
    return _spanList.SpanList;
  }
});
Object.defineProperty(exports, "maybeLoc", {
  enumerable: true,
  get: function () {
    return _spanList.maybeLoc;
  }
});
Object.defineProperty(exports, "loc", {
  enumerable: true,
  get: function () {
    return _spanList.loc;
  }
});
Object.defineProperty(exports, "hasSpan", {
  enumerable: true,
  get: function () {
    return _spanList.hasSpan;
  }
});
Object.defineProperty(exports, "node", {
  enumerable: true,
  get: function () {
    return _node.node;
  }
});
exports.ASTv2 = exports.AST = exports.ASTv1 = void 0;

var _source = require("./lib/source/source");

var _publicBuilders = _interopRequireDefault(require("./lib/v1/public-builders"));

var ASTv1_1 = _interopRequireWildcard(require("./lib/v1/api"));

var AST_1 = ASTv1_1;
exports.ASTv1 = ASTv1_1;
exports.AST = ASTv1_1;

var ASTv2_1 = _interopRequireWildcard(require("./lib/v2-a/api"));

exports.ASTv2 = ASTv2_1;

var _normalize = require("./lib/v2-a/normalize");

var _symbolTable = require("./lib/symbol-table");

var _syntaxError = require("./lib/syntax-error");

var _tokenizerEventHandlers = require("./lib/parser/tokenizer-event-handlers");

var _print = _interopRequireDefault(require("./lib/generation/print"));

var _util = require("./lib/generation/util");

var _walker = _interopRequireDefault(require("./lib/traversal/walker"));

var _traverse = _interopRequireDefault(require("./lib/traversal/traverse"));

var _errors = require("./lib/traversal/errors");

var _path = _interopRequireDefault(require("./lib/traversal/path"));

var _keywords = require("./lib/keywords");

var _getTemplateLocals = require("./lib/get-template-locals");

var _slice = require("./lib/source/slice");

var _span = require("./lib/source/span");

var _spanList = require("./lib/source/span-list");

var _node = require("./lib/v2-a/objects/node");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
