// @glimmer/syntax/dist/commonjs/es2017/lib/source/source.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Source = void 0;

var _env = require("@glimmer/env");

var _util = require("@glimmer/util");

var _span = require("./span");

// eslint-disable-next-line import/no-extraneous-dependencies
class Source {
  constructor(source, module = 'an unknown module') {
    this.source = source;
    this.module = module;
  }
  /**
   * Validate that the character offset represents a position in the source string.
   */


  check(offset) {
    return offset >= 0 && offset <= this.source.length;
  }

  slice(start, end) {
    return this.source.slice(start, end);
  }

  offsetFor(line, column) {
    return _span.SourceOffset.forHbsPos(this, {
      line,
      column
    });
  }

  spanFor({
    start,
    end
  }) {
    return _span.SourceSpan.forHbsLoc(this, {
      start: {
        line: start.line,
        column: start.column
      },
      end: {
        line: end.line,
        column: end.column
      }
    });
  }

  hbsPosFor(offset) {
    let seenLines = 0;
    let seenChars = 0;

    if (offset > this.source.length) {
      return null;
    }

    while (true) {
      let nextLine = this.source.indexOf('\n', seenChars);

      if (offset <= nextLine || nextLine === -1) {
        return {
          line: seenLines + 1,
          column: offset - seenChars
        };
      } else {
        seenLines += 1;
        seenChars = nextLine + 1;
      }
    }
  }

  charPosFor(position) {
    let {
      line,
      column
    } = position;
    let sourceString = this.source;
    let sourceLength = sourceString.length;
    let seenLines = 0;
    let seenChars = 0;

    while (true) {
      if (seenChars >= sourceLength) return sourceLength;
      let nextLine = this.source.indexOf('\n', seenChars);
      if (nextLine === -1) nextLine = this.source.length;

      if (seenLines === line - 1) {
        if (seenChars + column > nextLine) return nextLine;

        if (_env.DEBUG) {
          let roundTrip = this.hbsPosFor(seenChars + column);
          false && (0, _util.assert)(roundTrip !== null, `the returned offset failed to round-trip`);
          false && (0, _util.assert)(roundTrip.line === line, `the round-tripped line didn't match the original line`);
          false && (0, _util.assert)(roundTrip.column === column, `the round-tripped column didn't match the original column`);
        }

        return seenChars + column;
      } else if (nextLine === -1) {
        return 0;
      } else {
        seenLines += 1;
        seenChars = nextLine + 1;
      }
    }
  }

}

exports.Source = Source;
