// @glimmer/syntax/dist/commonjs/es2017/lib/source/span-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loc = loc;
exports.hasSpan = hasSpan;
exports.maybeLoc = maybeLoc;
exports.SpanList = void 0;

var _span = require("./span");

class SpanList {
  constructor(span = []) {
    this._span = span;
  }

  static range(span, fallback = _span.SourceSpan.NON_EXISTENT) {
    return new SpanList(span.map(loc)).getRangeOffset(fallback);
  }

  add(offset) {
    this._span.push(offset);
  }

  getRangeOffset(fallback) {
    if (this._span.length === 0) {
      return fallback;
    } else {
      let first = this._span[0];
      let last = this._span[this._span.length - 1];
      return first.extend(last);
    }
  }

}

exports.SpanList = SpanList;

function loc(span) {
  if (Array.isArray(span)) {
    let first = span[0];
    let last = span[span.length - 1];
    return loc(first).extend(loc(last));
  } else if (span instanceof _span.SourceSpan) {
    return span;
  } else {
    return span.loc;
  }
}

function hasSpan(span) {
  if (Array.isArray(span) && span.length === 0) {
    return false;
  }

  return true;
}

function maybeLoc(location, fallback) {
  if (hasSpan(location)) {
    return loc(location);
  } else {
    return fallback;
  }
}
