// @glimmer/util/dist/modules/es5/lib/platform-utils.js
import intern from './intern';
export var HAS_NATIVE_PROXY = typeof Proxy === 'function';
export var HAS_NATIVE_SYMBOL = function () {
  if (typeof Symbol !== 'function') {
    return false;
  } // eslint-disable-next-line symbol-description


  return typeof Symbol() === 'symbol';
}();
export function keys(obj) {
  return Object.keys(obj);
}
export function unwrap(val) {
  if (val === null || val === undefined) throw new Error("Expected value to be present");
  return val;
}
export function expect(val, message) {
  if (val === null || val === undefined) throw new Error(message);
  return val;
}
export function unreachable(message) {
  if (message === void 0) {
    message = 'unreachable';
  }

  return new Error(message);
}
export function exhausted(value) {
  throw new Error("Exhausted " + value);
}
export var tuple = function tuple() {
  for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return args;
};
export function enumerableSymbol(key) {
  return intern("__" + key + Math.floor(Math.random() * Date.now()) + "__");
}
export var symbol = HAS_NATIVE_SYMBOL ? Symbol : enumerableSymbol;
