// @glimmer/syntax/dist/commonjs/es2017/lib/v2-a/objects/attr-block.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ElementModifier = exports.ComponentArg = exports.SplatAttr = exports.HtmlAttr = void 0;

var _args = require("./args");

var _node = require("./node");

/**
 * `HtmlAttr` nodes are valid HTML attributes, with or without a value.
 *
 * Exceptions:
 *
 * - `...attributes` is `SplatAttr`
 * - `@x=<value>` is `ComponentArg`
 */
class HtmlAttr extends (0, _node.node)('HtmlAttr').fields() {}

exports.HtmlAttr = HtmlAttr;

class SplatAttr extends (0, _node.node)('SplatAttr').fields() {}
/**
 * Corresponds to an argument passed by a component (`@x=<value>`)
 */


exports.SplatAttr = SplatAttr;

class ComponentArg extends (0, _node.node)().fields() {
  /**
   * Convert the component argument into a named argument node
   */
  toNamedArgument() {
    return new _args.NamedArgument({
      name: this.name,
      value: this.value
    });
  }

}
/**
 * An `ElementModifier` is just a normal call node in modifier position.
 */


exports.ComponentArg = ComponentArg;

class ElementModifier extends (0, _node.node)('ElementModifier').fields() {}

exports.ElementModifier = ElementModifier;
