// @glimmer/syntax/dist/commonjs/es2017/lib/v2-a/objects/content.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SimpleElement = exports.InvokeComponent = exports.InvokeBlock = exports.AppendContent = exports.HtmlComment = exports.HtmlText = exports.GlimmerComment = void 0;

var _spanList = require("../../source/span-list");

var _args = require("./args");

var _node = require("./node");

class GlimmerComment extends (0, _node.node)('GlimmerComment').fields() {}

exports.GlimmerComment = GlimmerComment;

class HtmlText extends (0, _node.node)('HtmlText').fields() {}

exports.HtmlText = HtmlText;

class HtmlComment extends (0, _node.node)('HtmlComment').fields() {}

exports.HtmlComment = HtmlComment;

class AppendContent extends (0, _node.node)('AppendContent').fields() {
  get callee() {
    if (this.value.type === 'Call') {
      return this.value.callee;
    } else {
      return this.value;
    }
  }

  get args() {
    if (this.value.type === 'Call') {
      return this.value.args;
    } else {
      return _args.Args.empty(this.value.loc.collapse('end'));
    }
  }

}

exports.AppendContent = AppendContent;

class InvokeBlock extends (0, _node.node)('InvokeBlock').fields() {}
/**
 * Corresponds to a component invocation. When the content of a component invocation contains no
 * named blocks, `blocks` contains a single named block named `"default"`. When a component
 * invocation is self-closing, `blocks` is empty.
 */


exports.InvokeBlock = InvokeBlock;

class InvokeComponent extends (0, _node.node)('InvokeComponent').fields() {
  get args() {
    let entries = this.componentArgs.map(a => a.toNamedArgument());
    return _args.Args.named(new _args.NamedArguments({
      loc: _spanList.SpanList.range(entries, this.callee.loc.collapse('end')),
      entries
    }));
  }

}
/**
 * Corresponds to a simple HTML element. The AST allows component arguments and modifiers to support
 * future extensions.
 */


exports.InvokeComponent = InvokeComponent;

class SimpleElement extends (0, _node.node)('SimpleElement').fields() {
  get args() {
    let entries = this.componentArgs.map(a => a.toNamedArgument());
    return _args.Args.named(new _args.NamedArguments({
      loc: _spanList.SpanList.range(entries, this.tag.loc.collapse('end')),
      entries
    }));
  }

}

exports.SimpleElement = SimpleElement;
