// @glimmer/syntax/dist/commonjs/es2017/lib/v2-a/objects/expr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isLiteral = isLiteral;
exports.InterpolateExpression = exports.DeprecatedCallExpression = exports.CallExpression = exports.PathExpression = exports.LiteralExpression = void 0;

var _slice = require("../../source/slice");

var _node = require("./node");

/**
 * Corresponds to a Handlebars literal.
 *
 * @see {LiteralValue}
 */
class LiteralExpression extends (0, _node.node)('Literal').fields() {
  toSlice() {
    return new _slice.SourceSlice({
      loc: this.loc,
      chars: this.value
    });
  }

}
/**
 * Returns true if an input {@see ExpressionNode} is a literal.
 */


exports.LiteralExpression = LiteralExpression;

function isLiteral(node, kind) {
  if (node.type === 'Literal') {
    if (kind === undefined) {
      return true;
    } else if (kind === 'null') {
      return node.value === null;
    } else {
      return typeof node.value === kind;
    }
  } else {
    return false;
  }
}
/**
 * Corresponds to a path in expression position.
 *
 * ```hbs
 * this
 * this.x
 * @x
 * @x.y
 * x
 * x.y
 * ```
 */


class PathExpression extends (0, _node.node)('Path').fields() {}
/**
 * Corresponds to a parenthesized call expression.
 *
 * ```hbs
 * (x)
 * (x.y)
 * (x y)
 * (x.y z)
 * ```
 */


exports.PathExpression = PathExpression;

class CallExpression extends (0, _node.node)('Call').fields() {}
/**
 * Corresponds to a possible deprecated helper call. Must be:
 *
 * 1. A free variable (not this.foo, not @foo, not local).
 * 2. Argument-less.
 * 3. In a component invocation's named argument position.
 * 4. Not parenthesized (not @bar={{(helper)}}).
 * 5. Not interpolated (not @bar="{{helper}}").
 *
 * ```hbs
 * <Foo @bar={{helper}} />
 * ```
 */


exports.CallExpression = CallExpression;

class DeprecatedCallExpression extends (0, _node.node)('DeprecatedCall').fields() {}
/**
 * Corresponds to an interpolation in attribute value position.
 *
 * ```hbs
 * <a href="{{url}}.html"
 * ```
 */


exports.DeprecatedCallExpression = DeprecatedCallExpression;

class InterpolateExpression extends (0, _node.node)('Interpolate').fields() {}

exports.InterpolateExpression = InterpolateExpression;
