// @glimmer/syntax/dist/commonjs/es2017/lib/v2-a/objects/internal-node.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NamedBlock = exports.NamedBlocks = exports.Block = exports.Template = void 0;

var _spanList = require("../../source/span-list");

var _args = require("./args");

var _node = require("./node");

/**
 * Corresponds to an entire template.
 */
class Template extends (0, _node.node)().fields() {}
/**
 * Represents a block. In principle this could be merged with `NamedBlock`, because all cases
 * involving blocks have at least a notional name.
 */


exports.Template = Template;

class Block extends (0, _node.node)().fields() {}
/**
 * Corresponds to a collection of named blocks.
 */


exports.Block = Block;

class NamedBlocks extends (0, _node.node)().fields() {
  get(name) {
    return this.blocks.filter(block => block.name.chars === name)[0] || null;
  }

}
/**
 * Corresponds to a single named block. This is used for anonymous named blocks (`default` and
 * `else`).
 */


exports.NamedBlocks = NamedBlocks;

class NamedBlock extends (0, _node.node)().fields() {
  get args() {
    let entries = this.componentArgs.map(a => a.toNamedArgument());
    return _args.Args.named(new _args.NamedArguments({
      loc: _spanList.SpanList.range(entries, this.name.loc.collapse('end')),
      entries
    }));
  }

}

exports.NamedBlock = NamedBlock;
