// @glimmer/syntax/dist/commonjs/es5/lib/keywords.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isKeyword = isKeyword;
exports.KEYWORDS_TYPES = void 0;

function isKeyword(word) {
  return word in KEYWORDS_TYPES;
}
/**
 * This includes the full list of keywords currently in use in the template
 * language, and where their valid usages are.
 */


var KEYWORDS_TYPES = {
  component: ['Call', 'Append', 'Block'],
  "debugger": ['Append'],
  'each-in': ['Block'],
  each: ['Block'],
  'has-block-params': ['Call', 'Append'],
  'has-block': ['Call', 'Append'],
  helper: ['Call', 'Append'],
  "if": ['Call', 'Append', 'Block'],
  'in-element': ['Block'],
  "let": ['Block'],
  'link-to': ['Append', 'Block'],
  log: ['Call', 'Append'],
  modifier: ['Call'],
  mount: ['Append'],
  mut: ['Call', 'Append'],
  outlet: ['Append'],
  'query-params': ['Call'],
  readonly: ['Call', 'Append'],
  unbound: ['Call', 'Append'],
  unless: ['Call', 'Append', 'Block'],
  "with": ['Block'],
  "yield": ['Append']
};
exports.KEYWORDS_TYPES = KEYWORDS_TYPES;
