// @glimmer/syntax/dist/commonjs/es5/lib/source/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "NON_EXISTENT_LOCATION", {
  enumerable: true,
  get: function () {
    return _location.NON_EXISTENT_LOCATION;
  }
});
Object.defineProperty(exports, "UNKNOWN_POSITION", {
  enumerable: true,
  get: function () {
    return _location.UNKNOWN_POSITION;
  }
});
Object.defineProperty(exports, "SYNTHETIC_LOCATION", {
  enumerable: true,
  get: function () {
    return _location.SYNTHETIC_LOCATION;
  }
});
Object.defineProperty(exports, "Source", {
  enumerable: true,
  get: function () {
    return _source.Source;
  }
});
Object.defineProperty(exports, "maybeLoc", {
  enumerable: true,
  get: function () {
    return _spanList.maybeLoc;
  }
});
Object.defineProperty(exports, "SpanList", {
  enumerable: true,
  get: function () {
    return _spanList.SpanList;
  }
});
Object.defineProperty(exports, "hasSpan", {
  enumerable: true,
  get: function () {
    return _spanList.hasSpan;
  }
});
Object.defineProperty(exports, "loc", {
  enumerable: true,
  get: function () {
    return _spanList.loc;
  }
});

var _location = require("./location");

var _source = require("./source");

var _spanList = require("./span-list");
