// @glimmer/syntax/dist/commonjs/es5/lib/source/location.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isLocatedWithPositionsArray = isLocatedWithPositionsArray;
exports.isLocatedWithPositions = isLocatedWithPositions;
exports.BROKEN_LOCATION = exports.NON_EXISTENT_LOCATION = exports.TEMPORARY_LOCATION = exports.SYNTHETIC = exports.SYNTHETIC_LOCATION = exports.UNKNOWN_POSITION = void 0;

var _util = require("@glimmer/util");

var UNKNOWN_POSITION = Object.freeze({
  line: 1,
  column: 0
});
exports.UNKNOWN_POSITION = UNKNOWN_POSITION;
var SYNTHETIC_LOCATION = Object.freeze({
  source: '(synthetic)',
  start: UNKNOWN_POSITION,
  end: UNKNOWN_POSITION
});
/** @deprecated */

exports.SYNTHETIC_LOCATION = SYNTHETIC_LOCATION;
var SYNTHETIC = SYNTHETIC_LOCATION;
exports.SYNTHETIC = SYNTHETIC;
var TEMPORARY_LOCATION = Object.freeze({
  source: '(temporary)',
  start: UNKNOWN_POSITION,
  end: UNKNOWN_POSITION
});
exports.TEMPORARY_LOCATION = TEMPORARY_LOCATION;
var NON_EXISTENT_LOCATION = Object.freeze({
  source: '(nonexistent)',
  start: UNKNOWN_POSITION,
  end: UNKNOWN_POSITION
});
exports.NON_EXISTENT_LOCATION = NON_EXISTENT_LOCATION;
var BROKEN_LOCATION = Object.freeze({
  source: '(broken)',
  start: UNKNOWN_POSITION,
  end: UNKNOWN_POSITION
});
exports.BROKEN_LOCATION = BROKEN_LOCATION;

function isLocatedWithPositionsArray(location) {
  return (0, _util.isPresent)(location) && location.every(isLocatedWithPositions);
}

function isLocatedWithPositions(location) {
  return location.loc !== undefined;
}
