// @glimmer/syntax/dist/commonjs/es5/lib/source/span-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loc = loc;
exports.hasSpan = hasSpan;
exports.maybeLoc = maybeLoc;
exports.SpanList = void 0;

var _span = require("./span");

var SpanList = /*#__PURE__*/function () {
  function SpanList(span) {
    if (span === void 0) {
      span = [];
    }

    this._span = span;
  }

  SpanList.range = function range(span, fallback) {
    if (fallback === void 0) {
      fallback = _span.SourceSpan.NON_EXISTENT;
    }

    return new SpanList(span.map(loc)).getRangeOffset(fallback);
  };

  var _proto = SpanList.prototype;

  _proto.add = function add(offset) {
    this._span.push(offset);
  };

  _proto.getRangeOffset = function getRangeOffset(fallback) {
    if (this._span.length === 0) {
      return fallback;
    } else {
      var first = this._span[0];
      var last = this._span[this._span.length - 1];
      return first.extend(last);
    }
  };

  return SpanList;
}();

exports.SpanList = SpanList;

function loc(span) {
  if (Array.isArray(span)) {
    var first = span[0];
    var last = span[span.length - 1];
    return loc(first).extend(loc(last));
  } else if (span instanceof _span.SourceSpan) {
    return span;
  } else {
    return span.loc;
  }
}

function hasSpan(span) {
  if (Array.isArray(span) && span.length === 0) {
    return false;
  }

  return true;
}

function maybeLoc(location, fallback) {
  if (hasSpan(location)) {
    return loc(location);
  } else {
    return fallback;
  }
}
