// @glimmer/syntax/dist/modules/es2017/lib/v2-a/objects/content.js
import { SpanList } from '../../source/span-list';
import { Args, NamedArguments } from './args';
import { node } from './node';
export class GlimmerComment extends node('GlimmerComment').fields() {}
export class HtmlText extends node('HtmlText').fields() {}
export class HtmlComment extends node('HtmlComment').fields() {}
export class AppendContent extends node('AppendContent').fields() {
  get callee() {
    if (this.value.type === 'Call') {
      return this.value.callee;
    } else {
      return this.value;
    }
  }

  get args() {
    if (this.value.type === 'Call') {
      return this.value.args;
    } else {
      return Args.empty(this.value.loc.collapse('end'));
    }
  }

}
export class InvokeBlock extends node('InvokeBlock').fields() {}
/**
 * Corresponds to a component invocation. When the content of a component invocation contains no
 * named blocks, `blocks` contains a single named block named `"default"`. When a component
 * invocation is self-closing, `blocks` is empty.
 */

export class InvokeComponent extends node('InvokeComponent').fields() {
  get args() {
    let entries = this.componentArgs.map(a => a.toNamedArgument());
    return Args.named(new NamedArguments({
      loc: SpanList.range(entries, this.callee.loc.collapse('end')),
      entries
    }));
  }

}
/**
 * Corresponds to a simple HTML element. The AST allows component arguments and modifiers to support
 * future extensions.
 */

export class SimpleElement extends node('SimpleElement').fields() {
  get args() {
    let entries = this.componentArgs.map(a => a.toNamedArgument());
    return Args.named(new NamedArguments({
      loc: SpanList.range(entries, this.tag.loc.collapse('end')),
      entries
    }));
  }

}
