// @glimmer/syntax/dist/modules/es2017/lib/v2-a/objects/expr.js
import { SourceSlice } from '../../source/slice';
import { node } from './node';
/**
 * Corresponds to a Handlebars literal.
 *
 * @see {LiteralValue}
 */

export class LiteralExpression extends node('Literal').fields() {
  toSlice() {
    return new SourceSlice({
      loc: this.loc,
      chars: this.value
    });
  }

}
/**
 * Returns true if an input {@see ExpressionNode} is a literal.
 */

export function isLiteral(node, kind) {
  if (node.type === 'Literal') {
    if (kind === undefined) {
      return true;
    } else if (kind === 'null') {
      return node.value === null;
    } else {
      return typeof node.value === kind;
    }
  } else {
    return false;
  }
}
/**
 * Corresponds to a path in expression position.
 *
 * ```hbs
 * this
 * this.x
 * @x
 * @x.y
 * x
 * x.y
 * ```
 */

export class PathExpression extends node('Path').fields() {}
/**
 * Corresponds to a parenthesized call expression.
 *
 * ```hbs
 * (x)
 * (x.y)
 * (x y)
 * (x.y z)
 * ```
 */

export class CallExpression extends node('Call').fields() {}
/**
 * Corresponds to a possible deprecated helper call. Must be:
 *
 * 1. A free variable (not this.foo, not @foo, not local).
 * 2. Argument-less.
 * 3. In a component invocation's named argument position.
 * 4. Not parenthesized (not @bar={{(helper)}}).
 * 5. Not interpolated (not @bar="{{helper}}").
 *
 * ```hbs
 * <Foo @bar={{helper}} />
 * ```
 */

export class DeprecatedCallExpression extends node('DeprecatedCall').fields() {}
/**
 * Corresponds to an interpolation in attribute value position.
 *
 * ```hbs
 * <a href="{{url}}.html"
 * ```
 */

export class InterpolateExpression extends node('Interpolate').fields() {}
