// @glimmer/syntax/dist/modules/es2017/lib/v2-a/objects/internal-node.js
import { SpanList } from '../../source/span-list';
import { Args, NamedArguments } from './args';
import { node } from './node';
/**
 * Corresponds to an entire template.
 */

export class Template extends node().fields() {}
/**
 * Represents a block. In principle this could be merged with `NamedBlock`, because all cases
 * involving blocks have at least a notional name.
 */

export class Block extends node().fields() {}
/**
 * Corresponds to a collection of named blocks.
 */

export class NamedBlocks extends node().fields() {
  get(name) {
    return this.blocks.filter(block => block.name.chars === name)[0] || null;
  }

}
/**
 * Corresponds to a single named block. This is used for anonymous named blocks (`default` and
 * `else`).
 */

export class NamedBlock extends node().fields() {
  get args() {
    let entries = this.componentArgs.map(a => a.toNamedArgument());
    return Args.named(new NamedArguments({
      loc: SpanList.range(entries, this.name.loc.collapse('end')),
      entries
    }));
  }

}
