// @glimmer/syntax/dist/commonjs/es2017/lib/traversal/walker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

class Walker {
  constructor(order) {
    this.order = order;
    this.stack = [];
  }

  visit(node, callback) {
    if (!node) {
      return;
    }

    this.stack.push(node);

    if (this.order === 'post') {
      this.children(node, callback);
      callback(node, this);
    } else {
      callback(node, this);
      this.children(node, callback);
    }

    this.stack.pop();
  }

  children(node, callback) {
    switch (node.type) {
      case 'Block':
      case 'Template':
        return visitors.Program(this, node, callback);

      case 'ElementNode':
        return visitors.ElementNode(this, node, callback);

      case 'BlockStatement':
        return visitors.BlockStatement(this, node, callback);

      default:
        return;
    }
  }

}

exports.default = Walker;
const visitors = {
  Program(walker, node, callback) {
    for (let i = 0; i < node.body.length; i++) {
      walker.visit(node.body[i], callback);
    }
  },

  Template(walker, node, callback) {
    for (let i = 0; i < node.body.length; i++) {
      walker.visit(node.body[i], callback);
    }
  },

  Block(walker, node, callback) {
    for (let i = 0; i < node.body.length; i++) {
      walker.visit(node.body[i], callback);
    }
  },

  ElementNode(walker, node, callback) {
    for (let i = 0; i < node.children.length; i++) {
      walker.visit(node.children[i], callback);
    }
  },

  BlockStatement(walker, node, callback) {
    walker.visit(node.program, callback);
    walker.visit(node.inverse || null, callback);
  }

};
