// @glimmer/syntax/dist/modules/es5/lib/source/source.js
// eslint-disable-next-line import/no-extraneous-dependencies
import { DEBUG } from '@glimmer/env';
import { assert } from '@glimmer/util';
import { SourceOffset, SourceSpan } from './span';
export var Source = /*#__PURE__*/function () {
  function Source(source, module) {
    if (module === void 0) {
      module = 'an unknown module';
    }

    this.source = source;
    this.module = module;
  }
  /**
   * Validate that the character offset represents a position in the source string.
   */


  var _proto = Source.prototype;

  _proto.check = function check(offset) {
    return offset >= 0 && offset <= this.source.length;
  };

  _proto.slice = function slice(start, end) {
    return this.source.slice(start, end);
  };

  _proto.offsetFor = function offsetFor(line, column) {
    return SourceOffset.forHbsPos(this, {
      line: line,
      column: column
    });
  };

  _proto.spanFor = function spanFor(_ref) {
    var start = _ref.start,
        end = _ref.end;
    return SourceSpan.forHbsLoc(this, {
      start: {
        line: start.line,
        column: start.column
      },
      end: {
        line: end.line,
        column: end.column
      }
    });
  };

  _proto.hbsPosFor = function hbsPosFor(offset) {
    var seenLines = 0;
    var seenChars = 0;

    if (offset > this.source.length) {
      return null;
    }

    while (true) {
      var nextLine = this.source.indexOf('\n', seenChars);

      if (offset <= nextLine || nextLine === -1) {
        return {
          line: seenLines + 1,
          column: offset - seenChars
        };
      } else {
        seenLines += 1;
        seenChars = nextLine + 1;
      }
    }
  };

  _proto.charPosFor = function charPosFor(position) {
    var line = position.line,
        column = position.column;
    var sourceString = this.source;
    var sourceLength = sourceString.length;
    var seenLines = 0;
    var seenChars = 0;

    while (true) {
      if (seenChars >= sourceLength) return sourceLength;
      var nextLine = this.source.indexOf('\n', seenChars);
      if (nextLine === -1) nextLine = this.source.length;

      if (seenLines === line - 1) {
        if (seenChars + column > nextLine) return nextLine;

        if (DEBUG) {
          var roundTrip = this.hbsPosFor(seenChars + column);
          false && assert(roundTrip !== null, "the returned offset failed to round-trip");
          false && assert(roundTrip.line === line, "the round-tripped line didn't match the original line");
          false && assert(roundTrip.column === column, "the round-tripped column didn't match the original column");
        }

        return seenChars + column;
      } else if (nextLine === -1) {
        return 0;
      } else {
        seenLines += 1;
        seenChars = nextLine + 1;
      }
    }
  };

  return Source;
}();
